include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp

```diff
@@ -0,0 +1,1197 @@
+	Eigen::Matrix<double, 6, 6> &hessian,
+	PointCloudSource &trans_cloud,
+	Eigen::Matrix<double, 6, 1> &p,
+	bool compute_hessian)
+{
+	score_gradient.setZero();
+	hessian.setZero();
+	double score = 0;
+	int total_neighborhood_count = 0;
+  double nearest_voxel_score = 0;
+  size_t found_neigborhood_voxel_num = 0;
+
+  std::vector<double> scores(input_->points.size());
+  std::vector<double> nearest_voxel_scores(input_->points.size());
+  std::vector<size_t> found_neigborhood_voxel_nums(input_->points.size());
+  std::vector<Eigen::Matrix<double, 6, 1>, Eigen::aligned_allocator<Eigen::Matrix<double, 6, 1>>> score_gradients(input_->points.size());
+  std::vector<Eigen::Matrix<double, 6, 6>, Eigen::aligned_allocator<Eigen::Matrix<double, 6, 6>>> hessians(input_->points.size());
+  std::vector<int> neighborhood_counts(input_->points.size());
+  for (std::size_t i = 0; i < input_->points.size(); i++) {
+		scores[i] = 0;
+    nearest_voxel_scores[i] = 0;
+    found_neigborhood_voxel_nums[i] = 0;
+		score_gradients[i].setZero();
+		hessians[i].setZero();
+		neighborhood_counts[i] = 0;
+	}
+
+	// Precompute Angular Derivatives (eq. 6.19 and 6.21)[Magnusson 2009]
+	computeAngleDerivatives(p);
+
+  std::vector<std::vector<TargetGridLeafConstPtr>> neighborhoods(num_threads_);
+  std::vector<std::vector<float>> distancess(num_threads_);
+
+// 	// Update gradient and hessian for each point, line 17 in Algorithm 2 [Magnusson 2009]
+#pragma omp parallel for num_threads(num_threads_) schedule(guided, 8)
+	for (std::size_t idx = 0; idx < input_->points.size(); idx++)
+	{
+		int thread_n = omp_get_thread_num();
+
+		// Original Point and Transformed Point
+		PointSource x_pt, x_trans_pt;
+		// Original Point and Transformed Point (for math)
+		Eigen::Vector3d x, x_trans;
+		// Occupied Voxel
+		TargetGridLeafConstPtr cell;
+		// Inverse Covariance of Occupied Voxel
+		Eigen::Matrix3d c_inv;
+
+		// Initialize Point Gradient and Hessian
+		Eigen::Matrix<float, 4, 6> point_gradient_;
+		Eigen::Matrix<float, 24, 6> point_hessian_;
+		point_gradient_.setZero();
+		point_gradient_.block<3, 3>(0, 0).setIdentity();
+		point_hessian_.setZero();
+
+		x_trans_pt = trans_cloud.points[idx];
+
+		auto& neighborhood = neighborhoods[thread_n];
+		auto& distances = distancess[thread_n];
+
+		// // Find neighbors (Radius search has been experimentally faster than direct neighbor checking.
+		// switch (search_method) {
+		// case KDTREE:
+		// 	target_cells_.radiusSearch(x_trans_pt, resolution_, neighborhood, distances);
+		// 	break;
+		// case DIRECT26:
+		// 	target_cells_.getNeighborhoodAtPoint(x_trans_pt, neighborhood);
+		// 	break;
+		// default:
+		// case DIRECT7:
+		// 	target_cells_.getNeighborhoodAtPoint7(x_trans_pt, neighborhood);
+		// 	break;
+		// case DIRECT1:
+		// 	target_cells_.getNeighborhoodAtPoint1(x_trans_pt, neighborhood);
+		// 	break;
+		// }
```

Comment:
These comments are no longer relevant, but you won't delete it?

Reply:
I am mentioning this section.
![image](https://user-images.githubusercontent.com/24854875/211440056-9a3ef825-0102-4147-a3c8-016d2f8e61ef.png)

Reply:
Resolved: https://github.com/tier4/ndt_omp/pull/19/commits/24e52cd7cf998a6a44e36aad6fb8b9ec4150111a & https://github.com/tier4/ndt_omp/pull/19/commits/b29f6c41403a9ca593c46f488d9eb02691a76ff9